README.md

```diff
@@ -108,9 +108,28 @@ setup(cmdclass=cmdclass)
 - This package does not work with the deprecated `python setup.py bdist_wheel` or `python setup.py sdist` commands, PyPA recommends using the [build](https://pypa-build.readthedocs.io/en/latest/index.html) package (`pip install build && python -m build .`).
 - We recommend using `include_package_data=True` and `MANIFEST.in` to control the assets included in the [package](https://setuptools.readthedocs.io/en/latest/userguide/datafiles.html).
 - Tools like [`check-manifest`](https://github.com/mgedmin/check-manifest) or [`manifix`](https://github.com/vidartf/manifix) can be used to ensure the desired assets are included.
-- Simple uses of `data_files` can be handled in `setup.cfg` or in `setup.py`.  If recursive directories are needed use `get_data_files()` from this package.
+- Simple uses of `data_files` can be handled in `setup.cfg` or in `setup.py`.  For more advanced usage, see below paragraph.
 - Unfortunately `data_files` are not supported in `develop` mode (a limitation of `setuptools`).  You can work around it by doing a full install (`pip install .`) before the develop install (`pip install -e .`), or by adding a script to push the data files to `sys.base_prefix`.
 
+### Advanced Data Files Usage
+
+If the simple data_files support of `setup.cfg` is not sufficient for your needs, use `get_data_files()` from this package. The simplest use case is for including checked in files in a recursive (glob's `**` pattern) manner. In this case, set the `data_files` argument to the `setup()` function in `setup.py` to the output of the `get_data_files()` function.
+
+If you need to include files that are generated during the build, then add the following logic to the end of your `pre_dist` command:
+```python
+builder = npm_builder()
+def build_with_datafiles():
+    builder()
+    # needed as some dirs are generated as part of the build (or lazy):
+    <need reference to dist here!>.data_files = get_data_files([
```

Comment:
This is the issue. How do we dynamically update the data files here? Do we need to pass a reference of the `Command` to the builder function?

Reply:
This is called after the build step, which means the glob should work on files that are there after building.

Reply:
I'm confused. To be clear, I expect `get_data_files` to work here, but the question is what we do with that list of files? What attribute do we set to tell the build system "hey! add these files to the collection of data files" ?

Reply:
It gets passed to [`setuptools.setup()`](https://github.com/jupyterlab/jupyterlab/blob/886b2ab9999dd281d47e9970f60e7d3398e7020f/setup.py#L57).  I much prefer the way flit handles data files now: you give it a populated directory and it maps it right to the `data` directory in the wheel, see #128

Reply:
I just ran into this in IPython Parallel. The problem here is that the data files aren't known until _after_ `builder()` is called, and can _change_ if builder is called to regenerate files already present (see https://github.com/ipython/ipyparallel/issues/675). So the choice is:

1. run `builder()` unconditionally before `get_data_files()` before passing it to setup
2. run `builder()` in the normal fashion in `pre_dist` and then _modify_ `dist.data_filles`. This is why the deprecated implementations had the [handle_files](https://github.com/jupyter/jupyter-packaging/blob/fa95229a004d4c51f8e4c9593ad285e55a740b1c/jupyter_packaging/setupbase.py#L517-L536). This step is still required to produce correct output, but it has been removed.

It can be helped if these hooks were passed the distribution as an arg, rather than no args.

Reply:
Perhaps the simplest (and most flit-like) way to do it is instead of running the builder as a pre_dist, etc. step, run it at the top-level of setup.py (perhaps conditional on some sys.argv to avoid triggering it _too_ often).

That's what I ended up doing in https://github.com/ipython/ipyparallel/pull/680 with extra steps to avoid the unnecessary rebuilds that caused https://github.com/ipython/ipyparallel/issues/675.

- if generated files are not present: _always_ run the build and skip the custom command classes (e.g. install from git or git-archive)
- If they are present
  - and running from git: hook up command classes, since they may want rebuilds
  - and _not_ running from git (probably sdist): skip rebuild

Reply:
Thanks @minrk, I did something similar in https://github.com/jupyterlab/jupyterlab/pull/12208.  I agree we should make this cleaner and provide better guidance.